nixos/modules/services/misc/gitlab.nix

```diff
@@ -402,43 +506,113 @@ in {
       };
 
       secrets.secret = mkOption {
-        type = types.str;
+        type = with types; nullOr str;
+        default = null;
         description = ''
           The secret is used to encrypt variables in the DB. If
           you change or lose this key you will be unable to access variables
           stored in database.
 
           Make sure the secret is at least 30 characters and all random,
           no regular words or you'll be exposed to dictionary attacks.
+
+          This should not be used in production, since it's stored in
+          the world-readable nix store. Use
+          <option>services.gitlab.secrets.secretFile</option>
+          instead.
+        '';
+      };
+
+      secrets.secretFile = mkOption {
```

Comment:
With the `gitea` module secrets are automatically generated and stored in the `stateDir`, which a sysadmin is then free to change. Is there value in allowing the sysadmin to generate this key upfront? I would suggest this option should be removed entirely, and in the edge case that a sysadmin actually needs to specify the key they can go into `stateDir` and do so.

Reply:
Hm, yeah, I guess they could be generated on first run. This would however be less declarative and hide state that is essential to read any backups made of the gitlab instance, so while it's initially more convenient, it might be worse in the long run...

Reply:
Modules like this are intrinsically stateful. There is a database. There is a `stateDir`. If a sysadmin fails to backup any of the state this application generates/utilizes the application becomes useless.

I would suggest these keys are implementation details not worth bothering the sysadmin over.

Reply:
What I was hinting at was that the backups done using the rake tasks would backup everything _but_ the secrets. I'm guessing that's why they weren't just automatically generated like the `gitlab_shell_secret` is. To always generate them also seems a bit contradictory to what you want with `createLocally`, or do you just want nothing to be created at all if `createLocally` is false?

Reply:
I'll clarify I mean `services.gitea.database.createLocally` which is very separate than if you had a `services.gitea.createLocally` (I'm not sure what that might imply). A `database.createLocally` only implies that a local database and database user should be provisioned.

I'm not overly familiar with the `gitlab` module so if I'm wrong about automatically creating secrets please say so. Maybe the automated backups imply automatically generating secrets is a bad idea... though we didn't think that was the case with `gitea`.

Reply:
Yes, I got that, but at least two of the secrets are used to encrypt contents of the database. In a comment in another thread you said:

>Some people may wish to provision their own database, or be migrating from an existing install so `createLocally` is nice to have.

Which wouldn't be possible unless the secrets are also configurable.

Although I would agree that it's nice to just have to set `services.gitlab.enable = true` to get gitlab up and running, I feel like it would come at too high a cost in this case:

- Not offering the secrets as options makes migration and manual provision tricky and is potentially hostile to beginners who don't know of their importance
- Generating only unset secrets opens up a new can of worms of user errors (not that this was suggested, but I thought of it as an alternative)

What I can think of as a reasonable middle ground would be to introduce a new option `generateSecrets`, which would control this behavior, and have it turned off by default. Maybe this should also be done for gitea, erring on the side of control over magic, but I don't know enough about gitea to say. :)

Reply:
> it's nice to just have to set `services.gitlab.enable = true` to get gitlab up and running

That should generally be one of the top goals of all module authors. The beauty of NixOS lies right there :smile:

That being said... you raise some really good points. I'm still inclined to think that restoring backups might involve some work from the sysadmin (and every sysadmin is already taking backups of state regularly I sure hope) so I'm less concerned about the drawbacks than you are. As far as secrets options go I've said my piece and will defer to your judgement on the matter. Thanks for hearing me out :smile: